doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,412 @@
+else if 3 in supported_versions[openassetio_mediacreation.packageId()]:
+  mediacreation_version = 3
+else:
+  incompatible_with_manager_error()
+  return
+
+if 1 in supported_versions[companycorp_traits.packageId()]:
+  companycorp_version = 1
+else:
+  incompatible_with_manager_error()
+  return
+
+# Tell the manager up front what versions you are going to use.
+# These must be versions reported from `supportedSchemaVersions`.
+negotiated_schema_versions =
+{
+  openassetio_mediacreation.packageId():mediacreation_version,
+  companycorp_traits.packageId():companycorp_version
+}
+manager.informSchemaVersions(negotiated_schema_versions)
+
+... # Go on to use openassetio methods with trait views of the now
+    # agreed upon version for each trait package.
+```
+### Manager side
+```python
+def supportedSchemaVersions(this, hostSession):
+  # As a manager, we can likely directly declare which schema versions
+  # we support
+  return {
+   openassetio_mediacreation.packageId(), [1,2,3]
+   companycorp_traits.packageId(), [1]
+  }
+
+def informSchemaVersions(this, schemaVersions, hostSession):
+  # Record the schema versions that the host is going to use.
+  # We can use these to branch our internal functionality
+  this.__mediacreation_schema_version = schemaVersions[openassetio_mediacreation.packageId()]
+  this.__companycorp_schema_version = schemaVersions[companycorp_traits.packageId()]
+```
+
+### Pros
+- Decides on version up front, reducing low level trait by trait branching.
+- Versioning is technically optional, which may allow "best-effort" data
+  access. E.g. if the trait ID remains unchanged, and only new
+  properties are added, then a trait version bump is a non-breaking
+  change - `isImbued()` still returns `true` and older trait properties
+  are still accessible.
+- Specifications continue to work as they are currently conceived. If
+  both host and manager agree on the schema version, then the valid
+  selection of Specification view classes is trivially known.
+
+### Cons
+- Complex dance needs to be done up front, people may end up just
+  skipping it.
+- The potential to deliberately skip version negotiation to allow "best
+  effort" access patterns is very bug-prone, may be difficult to debug,
+  and in general introduces confusion about whether versions should be
+  accounted for.
+- Manager reporting a supported schema version could imply every trait
+  in that schema is supported, despite the need to still call
+  managementPolicy.
+- Managers and hosts now have additional data related to specific
+  entities that they must communicate, increasing the complexity of
+  "pass-through" hosts and managers, and complexity of data transfer
+  across multiple hops and/or storage in multiple systems. This could be
+  especially problematic for naive managers that simply wish to be data
+  recorders, as they are almost forced to be version aware.
+- Requires manager plugins to be stateful, at least at `initialze` time.
```

Comment:
[fix] Only if this is done as a pre-negotiated mechanism, nothing to stop you putting the versions list in the Context instead of `intialize`.

Reply:
I think we basically assume a negotiation mechanism, because simply communicating the supported versions from host to manager violates some requirements (as mentioned in the sentence just above the `### Negotiated` section)

Reply:
I guess it feels like all the options violate something somewhere so is it definitely worth ruling out that as an alternative tradeoff? ie if you dont like 'in-data' but also dont want to become stateful?

Reply:
I wouldn't rule it out, seems like a decent possibility to me. I swear we discussed it at some point ... but things have moved so much from then it was probably in a completely different ... no pun intended ... context.

Reply:
If we take this out, then statelessness can't be a pro of the in-data solution.

I've softened the language, as I think negotiation is pretty strongly implied by the other conditions, but admit that it's not a definite requirement to out-of-data. https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/6c01b0a09d88724bade5c0ebdba54eeeeb08b1b5